// react-imported-component/dist/es5/entrypoints/boot.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.addPreloader = exports.injectLoadableTracker = exports.setConfiguration = exports.loadByChunkname = exports.assignImportedComponents = exports.whenComponentsReady = exports.rehydrateMarks = void 0;
var config_1 = require("../configuration/config");
Object.defineProperty(exports, "setConfiguration", { enumerable: true, get: function () { return config_1.setConfiguration; } });
var assignImportedComponents_1 = require("../loadable/assignImportedComponents");
Object.defineProperty(exports, "assignImportedComponents", { enumerable: true, get: function () { return assignImportedComponents_1.assignImportedComponents; } });
var loadByChunkName_1 = require("../loadable/loadByChunkName");
Object.defineProperty(exports, "loadByChunkname", { enumerable: true, get: function () { return loadByChunkName_1.loadByChunkname; } });
var marks_1 = require("../loadable/marks");
Object.defineProperty(exports, "rehydrateMarks", { enumerable: true, get: function () { return marks_1.rehydrateMarks; } });
var pending_1 = require("../loadable/pending");
Object.defineProperty(exports, "whenComponentsReady", { enumerable: true, get: function () { return pending_1.done; } });
var preloaders_1 = require("../loadable/preloaders");
Object.defineProperty(exports, "addPreloader", { enumerable: true, get: function () { return preloaders_1.addPreloader; } });
var globalTracker_1 = require("../trackers/globalTracker");
Object.defineProperty(exports, "injectLoadableTracker", { enumerable: true, get: function () { return globalTracker_1.injectLoadableTracker; } });
