// react-imported-component/dist/es2019/entrypoints/macro.js
// @ts-ignore
import { createMacro } from 'babel-plugin-macros';
import { createTransformer } from '../babel/babel';
import { assignImportedComponents } from '../loadable/assignImportedComponents';
function getMacroType(tagName) {
    switch (tagName) {
        case 'importedModule':
        case 'imported':
        case 'lazy':
        case 'useImported':
        case 'ImportedModule':
        case 'ImportedComponent':
            return 'react-imported-component';
        case 'assignImportedComponents':
            return 'react-imported-component/boot';
        default:
            return false;
    }
}
function macro({ references, state, babel }) {
    const { types: t } = babel;
    const fileName = state.file.opts.filename;
    const imports = {};
    const transformer = createTransformer(babel, true);
    Object.keys(references).forEach((tagName) => {
        const origin = getMacroType(tagName);
        if (origin) {
            imports[origin] = imports[origin] || [];
            imports[origin].push(tagName);
            const tags = references[tagName];
            tags.forEach((tag) => {
                const expression = tag.parentPath;
                if (t.isCallExpression(expression)) {
                    transformer.traverse(expression, fileName);
                }
            });
        }
    });
    addReactImports(babel, state, imports);
    transformer.traverse(state.file.path, fileName);
    transformer.finish(state.file.path.node, fileName);
}
function addReactImports(babel, state, importsGroups) {
    const { types: t } = babel;
    return Object.keys(importsGroups)
        .map((origin) => {
        const imports = importsGroups[origin];
        if (!imports.length) {
            return false;
        }
        const importedImport = state.file.path.node.body.find((importNode) => t.isImportDeclaration(importNode) && importNode.source.value === origin);
        // Handle adding the import or altering the existing import
        if (importedImport) {
            imports.forEach((name) => {
                if (!importedImport.specifiers.find((specifier) => specifier.imported && specifier.imported.name === name)) {
                    importedImport.specifiers.push(t.importSpecifier(t.identifier(name), t.identifier(name)));
                }
            });
        }
        else {
            state.file.path.node.body.unshift(t.importDeclaration(imports.map((name) => t.importSpecifier(t.identifier(name), t.identifier(name))), t.stringLiteral(origin)));
        }
        return true;
    })
        .some(Boolean);
}
const neverCallMe = () => {
    throw new Error('you have used `react-imported-component/macro` without `babel-plugin-macro` or `react-hot-loader/babel` installed');
};
const lazy = neverCallMe;
const imported = neverCallMe;
const importedModule = neverCallMe;
const useImported = neverCallMe;
const ImportedModule = neverCallMe;
const ImportedComponent = neverCallMe;
export { lazy, imported, importedModule, ImportedModule, ImportedComponent, useImported, assignImportedComponents };
export default createMacro(macro);
