// react-imported-component/dist/es5/scanners/shared.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getFileContent = exports.pWriteFile = exports.pReadFile = exports.getMatchString = exports.getRelative = exports.normalizePath = void 0;
var fs_1 = require("fs");
var path_1 = require("path");
var util_1 = require("util");
var normalizePath = function (path) { return path.split(path_1.sep).join('/'); };
exports.normalizePath = normalizePath;
var getRelative = function (from, to) {
    // force one unit paths
    var rel = (0, exports.normalizePath)((0, path_1.relative)(from, to));
    return rel[0] !== '.' ? './' + rel : rel;
};
exports.getRelative = getRelative;
var getMatchString = function (pattern, selected) {
    return function (str) {
        return (str.match(new RegExp(pattern, 'g')) || []).map(function (statement) { return (statement.match(new RegExp(pattern, 'i')) || [])[selected]; });
    };
};
exports.getMatchString = getMatchString;
exports.pReadFile = (0, util_1.promisify)(fs_1.readFile);
exports.pWriteFile = (0, util_1.promisify)(fs_1.writeFile);
var getFileContent = function (file) { return (0, exports.pReadFile)(file, 'utf8'); };
exports.getFileContent = getFileContent;
